modules/xml/filterx-parse-windows-eventlog-xml.c

```diff
@@ -0,0 +1,277 @@
+               FilterXParseWindowsEventlogXmlUserData *user_data, GError **error)
+{
+  g_string_assign(user_data->last_data_name, attribute_values[0]);
+  user_data->has_named_data = TRUE;
+}
+
+static void
+_store_relevant_position(FilterXParseWindowsEventlogXmlUserData *user_data, const gchar *element_name)
+{
+  if (g_strcmp0(element_name, "Event") == 0)
+    user_data->in_event = TRUE;
+  else if (g_strcmp0(element_name, "EventData") == 0)
+    user_data->in_event_data = TRUE;
+  else if (g_strcmp0(element_name, "Data") == 0)
+    user_data->in_data = TRUE;
+}
+
+static gboolean
+_has_windows_eventlog_event_data_attr(const gchar **attribute_names, FilterXParseWindowsEventlogXmlUserData *user_data)
+{
+  if (!user_data->in_data)
+    return FALSE;
+
+  if (!attribute_names[0])
+    return FALSE;
+
+  if (g_strcmp0(attribute_names[0], "Name") != 0)
+    {
+      msg_debug("FilterX: parse_windows_eventlog_xml(): Unexpected attribute in Data, expected: Name",
+                evt_tag_str("attribute", attribute_names[0]));
+      return FALSE;
+    }
+
+  if (attribute_names[1])
+    {
+      msg_debug("FilterX: parse_windows_eventlog_xml(): Unexpected attribute in Data, number of attributes must be 1",
+                evt_tag_str("attribute", attribute_names[1]));
+      return FALSE;
+    }
+
+  return TRUE;
+}
+
+static void
+_start_elem(FilterXGeneratorFunctionParseXml *s,
+            GMarkupParseContext *context, const gchar *element_name,
+            const gchar **attribute_names, const gchar **attribute_values,
+            FilterXParseXmlUserData *u, GError **error)
+{
+  FilterXParseWindowsEventlogXmlUserData *user_data = (FilterXParseWindowsEventlogXmlUserData *) u;
+  XmlElemContext *last_elem_context = g_queue_peek_head(user_data->super.xml_elem_context_stack);
+
+  _store_relevant_position(user_data, element_name);
+
+  if (!_has_windows_eventlog_event_data_attr(attribute_names, user_data))
+    {
+      filterx_parse_xml_start_elem_method(s, context, element_name, attribute_names, attribute_values, u, error);
+      return;
+    }
+
+  if (!filterx_object_is_type(last_elem_context->current_obj, &FILTERX_TYPE_NAME(dict)))
+    {
+      if (!_convert_to_dict(context, last_elem_context, error))
+        return;
+    }
+
+  XmlElemContext *new_elem_context = _prepare_elem(element_name, last_elem_context, error);
+  if (!new_elem_context)
+    return;
+  g_queue_push_head(user_data->super.xml_elem_context_stack, new_elem_context);
+
+  _collect_attrs(attribute_names, attribute_values, user_data, error);
+}
+
+static void
+_end_elem(FilterXGeneratorFunctionParseXml *s,
+          GMarkupParseContext *context, const gchar *element_name,
+          FilterXParseXmlUserData *u, GError **error)
+{
+  FilterXParseWindowsEventlogXmlUserData *user_data = (FilterXParseWindowsEventlogXmlUserData *) u;
+
+  if (user_data->in_data && g_strcmp0(element_name, "Data") == 0)
```

Comment:
this list of items is pretty close to what we have on the opening tag (e.g. _store_relevant_position()), we could potentially use the same function, with an explicit value instead of a hard-coded TRUE or FALSE.

Reply:
Yes, but you need to have a reverse order of checking on the closing part, at least I could only make it work that way. Maybe some refactor of this small part could make it work, I'll think about it.

Reply:
ah, ok. I didn't notice the different order. Also the starting part does not check that the booleans are false. Isn't that a problem?

Reply:
I'm rewriting this part with an enum and a very basic state machine, I think it is cleaner. That will fix the false checking problem.

Reply:
Fixed.